src/server/worker/utils/utils.go

```diff
@@ -0,0 +1,662 @@
+	"gopkg.in/src-d/go-git.v4"
+	gitPlumbing "gopkg.in/src-d/go-git.v4/plumbing"
+
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/exec"
+	"github.com/pachyderm/pachyderm/src/server/pkg/hashtree"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	filesync "github.com/pachyderm/pachyderm/src/server/pkg/sync"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/worker/common"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+	"github.com/pachyderm/pachyderm/src/server/worker/stats"
+)
+
+const (
+	// The maximum number of concurrent download/upload operations
+	concurrency = 100
+
+	chunkPrefix = "/chunk"
+	mergePrefix = "/merge"
+)
+
+type Utils interface {
+	Jobs() col.Collection
+	Pipelines() col.Collection
+	Chunks(jobID string) col.Collection
+	Merges(jobID string) col.Collection
+
+	WithProvisionedNode(context.Context, logs.TaggedLogger, func(*pps.ProcessStats) error, []*common.Input, *hashtree.Ordered) (*pps.ProcessStats, error)
+	RunUserCode(context.Context, logs.TaggedLogger, []string, *pps.ProcessStats, *types.Duration) error
+	RunUserErrorHandlingCode(context.Context, logs.TaggedLogger, []string, *pps.ProcessStats, *types.Duration) error
+
+	DeleteJob(col.STM, *pps.EtcdJobInfo) error
+	UpdateJobState(context.Context, string, *pfs.Commit, pps.JobState, string) error
+}
+
+type utils struct {
+	pipelineInfo *pps.PipelineInfo
+	pachClient   *client.APIClient
+	etcdClient   *etcd.Client
+	etcdPrefix   string
+	jobs         col.Collection
+	pipelines    col.Collection
+
+	uid *uint32
+	gid *uint32
+}
+
+func NewUtils(pipelineInfo *pps.PipelineInfo, pachClient *client.APIClient, etcdClient *etcd.Client, etcdPrefix string) (Utils, error) {
+	result := &utils{
+		pipelineInfo: pipelineInfo,
+		pachClient:   pachClient,
+		etcdClient:   etcdClient,
+		etcdPrefix:   etcdPrefix,
+		jobs:         ppsdb.Jobs(etcdClient, etcdPrefix),
+		pipelines:    ppsdb.Pipelines(etcdClient, etcdPrefix),
+	}
+
+	if pipelineInfo.Transform.User != "" {
+		user, err := lookupDockerUser(pipelineInfo.Transform.User)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
+		}
+		// If `user` is `nil`, `uid` and `gid` will get set, and we won't
+		// customize the user that executes the worker process.
+		if user != nil { // user is nil when os.IsNotExist(err) is true in which case we use root
+			uid, err := strconv.ParseUint(user.Uid, 10, 32)
+			if err != nil {
+				return nil, err
+			}
+			uid32 := uint32(uid)
+			result.uid = &uid32
+			gid, err := strconv.ParseUint(user.Gid, 10, 32)
+			if err != nil {
+				return nil, err
+			}
+			gid32 := uint32(gid)
+			result.gid = &gid32
+		}
+	}
+
+	return result, nil
+}
+
```

Comment:
How closely should I review the functions below? It looks like they were all copied over from other places and thus there's not much point in taking a close look at them, but if there's something that was added or changed I'm happy to take a look.

Reply:
I'll update the base branch for all functions that I moved so that you only see the changes I made on top of that.